cmd/clair/httpauth.go

```diff
@@ -0,0 +1,215 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gregjones/httpcache"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// AuthCheck is an interface that reports whether the passed request should be
+// allowed to continue.
+type AuthCheck interface {
+	Check(context.Context, *http.Request) bool
+}
+
+type authHandler struct {
+	auth AuthCheck
+	next http.Handler
+}
+
+func (h *authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !h.auth.Check(r.Context(), r) {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+	h.next.ServeHTTP(w, r)
+}
+
+// Authorize returns a Handler that gates access to the passed Handler behind
+// the passed AuthCheck.
+func Authorize(h http.Handler, f AuthCheck) http.Handler {
+	return &authHandler{
+		auth: f,
+		next: h,
+	}
+}
+
+func fromHeader(r *http.Request) (string, bool) {
+	hs, ok := r.Header["Authorization"]
+	if !ok {
+		return "", false
+	}
+	for _, h := range hs {
+		if strings.HasPrefix(h, "Bearer ") {
+			return strings.TrimPrefix(h, "Bearer "), true
+		}
+	}
+	return "", false
+}
+
+// PSKAuth returns an AuthCheck that validates a JWT with the supplied key and
+// ensures the issuer claim matches.
+func PSKAuth(key []byte, issuer string) (AuthCheck, error) {
+	return &psk{
+		key: key,
+		iss: issuer,
+	}, nil
+}
+
+type psk struct {
+	key []byte
+	iss string
+}
+
+func (p *psk) Check(_ context.Context, r *http.Request) bool {
+	wt, ok := fromHeader(r)
+	if !ok {
+		return false
+	}
+	tok, err := jwt.ParseSigned(wt)
+	if err != nil {
+		return false
+	}
+	if tok.Headers[0].Algorithm != string(jose.HS256) {
+		return false
+	}
+	cl := jwt.Claims{}
+	if err := tok.Claims(p.key, &cl); err != nil {
+		return false
+	}
+	if err := cl.ValidateWithLeeway(jwt.Expected{
+		Issuer: p.iss,
+		Time:   time.Now(),
+	}, 15*time.Second); err != nil {
+		return false
+	}
+	return true
+}
+
+type ks struct {
+	root   *url.URL
+	client *http.Client
+	mu     sync.RWMutex
+	cache  map[string]*jose.JSONWebKey
+}
+
+// Check implments AuthCheck.
+func (s *ks) Check(ctx context.Context, r *http.Request) bool {
+	wt, ok := fromHeader(r)
+	if !ok {
+		return false
+	}
+	tok, err := jwt.ParseSigned(wt)
+	if err != nil {
+		return false
+	}
+	aud, err := r.URL.Parse("/")
+	if err != nil {
+		return false
+	}
+	// Need to find a the key id.
+	ok = false
+	var kid string
+	for _, h := range tok.Headers {
+		if h.Algorithm == string(jose.RS256) {
+			ok = true
+			kid = h.KeyID
+			break
+		}
+	}
+	if !ok {
+		return false
+	}
+	// Need to pull out the issuer to fetch the key. We cannot return "true"
+	// until *after* a keyed Claims call succeeds.
```

Comment:
what is "keyed Claims call" ?

Reply:
The call under this is unsafe, so we can't trust it until we fetch the key and make the normal Claims call with it.